### ISSUE TITLE
Implement Code Splitting

### ISSUE THREAD

**Issue by leoloso** [leoloso/PoP#54]
The idea is to initially load assets as small as they can be, containing only what is needed for that page (JS, CSS, Handlebars templates) and nothing more. This way, loading the first page will get a significant performance boost.

The challenge is how to make it work with the "previous session openTabs()" functionality, which allows the user to immediately open many more pages to the current one, possibly from pages cached using Service Workers. This implies that, by the time these pages from the Service Workers cache are loaded (which is pretty much immediately), their assets must also have been pre-loaded. 

Can that be achieved using code-splitting? If so, will the code be manageable? (eg: importing all the Handlebars templates one per file is not pretty... it's normally nicer to import them as a bundle.) Does it make sense to implement it? 

As a reference, WebPack is doing [code-splitting](https://webpack.js.org/guides/code-splitting/#resource-splitting)

**Comment by leoloso** [comment]
Code splitting

The objective of code splitting is to reduce the amount of javascript sent to the page, so that the JS engine's compilation and parsing time goes down. It is this time that makes the app become slow in mobile phones, so shipping less javascript code will increase the time to interactive.

There are 2 places where we can do this: 1. on the Handlebars templates bundle file, and 2. on the JS bundle file.

### 1. Handlebars templates bundle file

A simple strategy can work out for the Handlebars templates bundle file: each configuration page knows what templates it will need (visible under `popManager.memory.settings['template-sources']` ), so then we can easily, for that configuration page, divide the bundle file into 2: initial (including templates which need to be loaded before rendering the page) and later (including those non-needed templates for that page, so they can be loaded later on). These 2 files will be loaded when rendering the initial page, and then never again (SPA kicks in).

How do we generate these 2 bundled files? It can be done on runtime:

- When loading a page, which has a certain configuration (over whose name we apply a hash function to generate {$TEMPLATE_BUNDLE}), check if the 2 bundle files exist in wp-content/pop-codesplitting/{$TEMPLATE_BUNDLE}-initial.js and {$TEMPLATE_BUNDLE}-later.js
- If they don't, then we collect all the templates from 'template-sources', and concatenate them all together into {$TEMPLATE_BUNDLE}-initial.js
- We have the list of all the existing templates for the app (under PoP_ServerSideRendering::$templatesource_paths), so we concatenate all the remaining templates into {$TEMPLATE_BUNDLE}-later.js
- We then include the initial bundle before calling `popManager.init()`, and the later bundle after

Pay attention to execute openSessionTabs() only after including all files!

### 2. JS bundle file

We can split all javascript code in the website into 2: 1. framework, which is static and always the same, containing code by the PoP framework that must be executed when loading the app, and 2. application, which contains the implementation code for that website. The application code can also be divided into 2: {$JS_BUNDLE}-initial.js and {$JS_BUNDLE}-later.js, containing code needed for that specific loaded page, and everything else. Then, instead of loading the whole js code, in advance we only load the framework file and {$JS_BUNDLE}-initial.js, and after doing `popManager.init()` we load {$JS_BUNDLE}-later.js.

The framework file is static, and it's always the same, so we simply must analyze what code goes there, extract it, and that's it.

All the rest is the application code, from which we must generate the 2 bundled files, initial and later.js. How to do this?

Still thinking...

**Comment by leoloso** [comment]
There is another possibility: we can calculate what resources (.js and handlebars .js) are needed for each page configuration by page configuration, print this information on the initial request (`loading_frame()`), and then when clicking on a link, inspect the pattern of the URL, identify to which configuration page it belongs, and load those needed resources immediately, even before the request is back from the server.

**Comment by leoloso** [comment]
We must find a way to tell, for each page configuration, what .js files it needs to request. For handlebars, it is straightforward. For .js functions, not so, so we gotta find out how to do it.

We already know what functions will be invoked. Then, for each of these functions, on a file by file basis, we create a configuration file which tells what are their dependencies, as in function name from what file. Then, knowing the initial input, we can generate the tree of dependencies. Then, we load the whole file, so each .js file may have only 1 function (or 1 set of inter-dependent functions). 

Loading all the portions of all files combined into 1 (generated from the build process for each page configuration) will not work, because:
- it is too complex (at least for the time being)
- we will be loading time and again the same functions, which appear in different bundles

Instead, simply loading the whole file, allows us to not re-load that file if it had already been loaded by another configuration page.

### How do we create the map of dependencies from .js files/functions?

- To create the starting functions: can read it directly from the .js file? (Eg: parsing and extracting the functions from "popJSLibraryManager.register(...)" in all .js files)
- From there, create a configuration file with the dependencies for each function. In PHP? XML? YAML? Any solution already done for this?

**Comment by leoloso** [comment]
Creating the configuration file with the dependencies for each function should be enough, it should already provide the starting point for generating the dependencies map. So there is no need to parse .js files for "popJSLibraryManager.register(...)"

**Comment by leoloso** [comment]
Implemented
